src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/MonitorConnectionContext.java

```diff
@@ -81,28 +82,47 @@ void setFailureCount(int failureCount) {
     this.failureCount = failureCount;
   }
 
+  void setInvalidNodeStartTime(long invalidNodeStartTimeMillis) {
+    this.invalidNodeStartTime = invalidNodeStartTimeMillis;
+  }
+
+  void resetInvalidNodeStartTime() { this.invalidNodeStartTime = 0; }
+
+  boolean isInvalidNodeStartTimeDefined() { return this.invalidNodeStartTime > 0; }
+
+  public long getInvalidNodeStartTime() { return this.invalidNodeStartTime; }
+
   public boolean isNodeUnhealthy() {
     return this.isNodeUnhealthy;
   }
 
-  void updateConnectionStatus(long currentTime, boolean isValid, long connectionValidationElapsedTime) {
+  void updateConnectionStatus(long currentTime, boolean isValid, long validationIntervalTimeMillis) {
     final long totalElapsedTimeMillis = currentTime - this.startMonitorTime;
 
     if (totalElapsedTimeMillis > this.failureDetectionTimeMillis) {
-      this.setConnectionValid(isValid && (this.failureDetectionIntervalMillis >= connectionValidationElapsedTime));
+      this.setConnectionValid(isValid, currentTime, validationIntervalTimeMillis);
     }
   }
 
-  void setConnectionValid(boolean connectionValid) {
+  void setConnectionValid(boolean connectionValid, long currentTime, long validationIntervalTimeMillis) {
     if (!connectionValid) {
       this.failureCount++;
 
-      if (this.getFailureCount() >= this.getFailureDetectionCount()) {
+      if(!this.isInvalidNodeStartTimeDefined()) {
+        this.setInvalidNodeStartTime(currentTime);
+      }
+
+      long invalidNodeDurationMillis = currentTime - this.getInvalidNodeStartTime();
+      long maxInvalidNodeDurationMillis = this.getFailureDetectionIntervalMillis() * this.getFailureDetectionCount();
+      float adjustedFailureCount = this.getFailureDetectionIntervalMillis() / validationIntervalTimeMillis * this.getFailureDetectionCount();
```

Comment:
In this division we used `float` in case we get some very small numbers, but I am not sure if this works.

If we obscure away all the names, we are essentially doing this:
`long adjustedFailureCount = int / long * int`
if `long` is very large, we are still going to get 0 with `float`.

Example:
```Java
final int aInt = 300;
final int bInt = 500;
final long cLong = Integer.MAX_VALUE;

float adjustedFailureCount1 = aInt / cLong * bInt;
System.out.println(adjustedFailureCount1);
// outputs: 0.0
```
We have to use `double` here instead.
```Java
final int aInt = 300;
final int bInt = 500;
final long cLong = Integer.MAX_VALUE;

float adjustedFailureCount1 = aInt / cLong * bInt;
System.out.println("adjustedFailureCount1: " + adjustedFailureCount1);

double adjustedFailureCount2 = (double) aInt / cLong * bInt;
System.out.println("adjustedFailureCount2: " + adjustedFailureCount2);

// outputs:
// adjustedFailureCount1: 0.0
// adjustedFailureCount2: 6.984919312868695E-5
```